pallet-chainlink-feed/src/lib.rs

```diff
@@ -0,0 +1,1408 @@
+	type MinimumReserve: Get<BalanceOf<Self>>;
+
+	/// Maximum allowed string length.
+	type StringLimit: Get<u32>;
+
+	/// Maximum number of oracles per feed.
+	type OracleCountLimit: Get<u32>;
+
+	/// Maximum number of feeds.
+	type FeedLimit: Get<Self::FeedId>;
+
+	/// Number of rounds to keep around per feed.
+	type PruningWindow: Get<RoundId>;
+
+	/// The weight for this pallet's extrinsics.
+	type WeightInfo: WeightInfo;
+}
+
+/// The configuration for an oracle feed.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct FeedConfig<
+	AccountId: Parameter,
+	Balance: Parameter,
+	BlockNumber: Parameter,
+	Value: Parameter,
+> {
+	owner: AccountId,
+	pending_owner: Option<AccountId>,
+	submission_value_bounds: (Value, Value),
+	submission_count_bounds: (u32, u32),
+	payment: Balance,
+	timeout: BlockNumber,
+	decimals: u8,
+	description: Vec<u8>,
+	restart_delay: RoundId,
+	reporting_round: RoundId,
+	latest_round: RoundId,
+	first_valid_round: Option<RoundId>,
+	oracle_count: u32,
+}
+pub type FeedConfigOf<T> = FeedConfig<
+	<T as frame_system::Trait>::AccountId,
+	BalanceOf<T>,
+	<T as frame_system::Trait>::BlockNumber,
+	<T as Trait>::Value,
+>;
+
+/// Round data relevant to consumers.
+/// Will only be constructed once minimum amount of submissions have
+/// been provided.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct Round<BlockNumber, Value> {
+	started_at: BlockNumber,
+	answer: Option<Value>,
+	updated_at: Option<BlockNumber>,
+	answered_in_round: Option<RoundId>,
+}
+pub type RoundOf<T> = Round<<T as frame_system::Trait>::BlockNumber, <T as Trait>::Value>;
+
+impl<BlockNumber, Value> Round<BlockNumber, Value>
+where
+	BlockNumber: Default, // BlockNumber
+	Value: Default,       // Value
+{
+	/// Create a new Round with the given starting block.
+	fn new(started_at: BlockNumber) -> Self {
+		Self {
+			started_at,
+			..Default::default()
+		}
+	}
+}
+
+/// Round data relevant to oracles.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct RoundDetails<Balance, BlockNumber, Value> {
+	submissions: Vec<Value>,
+	submission_count_bounds: (u32, u32),
+	payment: Balance,
+	timeout: BlockNumber,
+}
+pub type RoundDetailsOf<T> =
+	RoundDetails<BalanceOf<T>, <T as frame_system::Trait>::BlockNumber, <T as Trait>::Value>;
+
+/// Meta data tracking withdrawable rewards and admin for an oracle.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct OracleMeta<AccountId, Balance> {
+	withdrawable: Balance,
+	admin: AccountId,
+	pending_admin: Option<AccountId>,
+}
+pub type OracleMetaOf<T> = OracleMeta<<T as frame_system::Trait>::AccountId, BalanceOf<T>>;
+
+/// Meta data tracking the oracle status for a feed.
+#[derive(Clone, Encode, Decode, Default, Eq, PartialEq, RuntimeDebug)]
+pub struct OracleStatus<Value> {
+	starting_round: RoundId,
+	ending_round: Option<RoundId>,
```

Comment:
Is `ending_round` being used consistently in relation with the concept of `disabled oracle`? Sometimes `enabled` means only `Some` and sometimes also `ending_round >= round_id`. (but maybe I'm missing something)

Reply:
It is consistently meant to track whether an oracle is enabled, yes.
If `ending_round` is < than `round_id` then that means the oracle is disabled for that feed.